File: backend/src/main/java/com/example/BACKEND/catalogue/decision/presentation/executive/SemanticMetricFormatter.java
package com.example.BACKEND.catalogue.decision.presentation.executive;

import org.springframework.stereotype.Component;

import java.util.Locale;
import java.util.Set;

/**
 * Single source of truth for executive number formatting (KPIs, tables, charts).
 */
@Component
public class SemanticMetricFormatter {

    private static final Set<String> KNOWN_FORMATS =
            Set.of("currency", "percent", "number", "text", "date");

    public String asMultiple(double ratio) {
        if (ratio <= 0 || Double.isNaN(ratio) || Double.isInfinite(ratio)) return "";
        if (ratio >= 10) return String.format(Locale.ROOT, "%.0f×", ratio);
        if (ratio >= 2) return String.format(Locale.ROOT, "%.1f×", ratio);
        return String.format(Locale.ROOT, "%.2f×", ratio);
    }

    /** Converts a percent delta (e.g. 750) into a relative multiple phrase fragment (8.5×). */
    public String percentDeltaAsMultiple(double deltaPct) {
        double multiple = 1.0 + (Math.abs(deltaPct) / 100.0);
        return asMultiple(multiple);
    }

    public String asSharePct(double pct) {
        if (PresentationValueSanitizer.isUnavailable(pct)) {
            return PresentationValueSanitizer.DASH;
        }
        if (pct >= 100) return String.format(Locale.ROOT, "%.0f%%", pct);
        return String.format(Locale.ROOT, "%.1f%%", pct);
    }

    public String asCurrency(double amount) {
        if (PresentationValueSanitizer.isUnavailable(amount)) {
            return PresentationValueSanitizer.DASH;
        }
        double abs = Math.abs(amount);
        String sign = amount < 0 ? "-" : "";
        if (abs >= 1_000_000_000) {
            double b = abs / 1_000_000_000;
            return sign + "$" + String.format(Locale.ROOT, b >= 10 ? "%.1fB" : "%.2fB", b);
        }
        if (abs >= 1_000_000) {
            double m = abs / 1_000_000;
            return sign + "$" + String.format(Locale.ROOT, m >= 100 ? "%.0fM" : "%.1fM", m);
        }
        if (abs >= 1_000) {
            return sign + "$" + String.format(Locale.ROOT, "%.1fK", abs / 1_000);
        }
        if (abs >= 100) {
            return sign + "$" + String.format(Locale.ROOT, "%.0f", abs);
        }
        return sign + "$" + String.format(Locale.ROOT, "%.2f", abs);
    }

    public String formatPercentValue(double value) {
        if (PresentationValueSanitizer.isUnavailable(value)) {
            return PresentationValueSanitizer.DASH;
        }
        if (value > 0 && value <= 1.0) {
            return asSharePct(value * 100.0);
        }
        return asSharePct(value);
    }

    public String detectFormat(String columnKey) {
        if (columnKey == null || columnKey.isBlank()) {
            return "number";
        }
        if (KNOWN_FORMATS.contains(columnKey.toLowerCase(Locale.ROOT))) {
            return columnKey.toLowerCase(Locale.ROOT);
        }
        String key = columnKey.toLowerCase(Locale.ROOT);
        if (key.contains("date") || key.contains("time") || key.contains("timestamp")
                || key.endsWith("_day") || key.endsWith("_month") || key.endsWith("_year")) {
            return "date";
        }
        if (key.contains("revenue") || key.contains("amount") || key.contains("fare")
                || key.contains("cost") || key.contains("price") || key.contains("fee")
                || key.contains("tip") || key.contains("total") || key.contains("sales")) {
            return "currency";
        }
        if (key.contains("pct") || key.contains("percent") || key.contains("share")
                || key.contains("rate") || key.contains("_ratio") || key.endsWith("ratio")) {
            return "percent";
        }
        if (key.equals("count") || key.endsWith("_count") || key.startsWith("count_")
                || key.contains("id") || key.contains("rank")
                || key.contains("coefficient") || key.contains("distance")
                || key.contains("hours") || key.contains("quantity")) {
            return "number";
        }
        return "text";
    }

    public String formatForDisplay(double value, String formatOrColumnKey) {
        if (Double.isNaN(value) || Double.isInfinite(value)) {
            return "—";
        }
        String format = KNOWN_FORMATS.contains(formatOrColumnKey != null
                ? formatOrColumnKey.toLowerCase(Locale.ROOT) : "")
                ? formatOrColumnKey.toLowerCase(Locale.ROOT)
                : detectFormat(formatOrColumnKey);
        return switch (format) {
            case "currency" -> asCurrency(value);
            case "percent" -> formatPercentValue(value);
            case "number" -> compactNumber(value);
            default -> compactNumber(value);
        };
    }

    public String formatValue(double value, String metricKey) {
        return formatForDisplay(value, metricKey);
    }

    public String compactNumber(double n) {
        if (PresentationValueSanitizer.isUnavailable(n)) {
            return PresentationValueSanitizer.DASH;
        }
        double abs = Math.abs(n);
        if (abs >= 1_000_000_000) return String.format(Locale.ROOT, "%.1fB", n / 1_000_000_000);
        if (abs >= 1_000_000) return String.format(Locale.ROOT, "%.0fM", n / 1_000_000);
        if (abs >= 1_000) return String.format(Locale.ROOT, "%.1fK", n / 1_000);
        if (abs >= 100 || abs == Math.rint(abs)) return String.format(Locale.ROOT, "%.0f", n);
        return String.format(Locale.ROOT, "%.1f", n);
    }
}

File: frontend/src/utils/semanticMetricFormatter.js
/**
 * Shared business number formatter — mirrors backend SemanticMetricFormatter
 * and presentation-layer detectFormat rules.
 */

import { DASH, NOT_AVAILABLE } from './presentationSanitizer';

const KNOWN_FORMATS = new Set(['currency', 'percent', 'number', 'text', 'date']);

export function detectFormat(columnKey) {
  if (!columnKey || typeof columnKey !== 'string') return 'number';
  const key = columnKey.toLowerCase();

  if (key.includes('date') || key.includes('time') || key.includes('timestamp')
      || key.endsWith('_day') || key.endsWith('_month') || key.endsWith('_year')) {
    return 'date';
  }
  if (key.includes('revenue') || key.includes('amount') || key.includes('fare')
      || key.includes('cost') || key.includes('price') || key.includes('fee')
      || key.includes('tip') || key.includes('total') || key.includes('sales')) {
    return 'currency';
  }
  if (key.includes('pct') || key.includes('percent') || key.includes('share')
      || key.includes('rate') || key.includes('_ratio') || key.endsWith('ratio')) {
    return 'percent';
  }
  if (key === 'count' || key.endsWith('_count') || key.startsWith('count_')
      || key.includes('id') || key.includes('rank')
      || key.includes('coefficient') || key.includes('distance')
      || key.includes('hours') || key.includes('quantity')) {
    return 'number';
  }
  return 'text';
}

export function asCurrency(amount) {
  const n = Number(amount);
  if (!Number.isFinite(n)) return '—';
  const abs = Math.abs(n);
  const sign = n < 0 ? '-' : '';
  if (abs >= 1_000_000_000) {
    const b = abs / 1_000_000_000;
    return `${sign}$${b >= 10 ? b.toFixed(1) : b.toFixed(2)}B`;
  }
  if (abs >= 1_000_000) {
    const m = abs / 1_000_000;
    return `${sign}$${m >= 100 ? m.toFixed(0) : m.toFixed(1)}M`;
  }
  if (abs >= 1_000) {
    return `${sign}$${(abs / 1_000).toFixed(1)}K`;
  }
  if (abs >= 100) {
    return `${sign}$${abs.toFixed(0)}`;
  }
  return `${sign}$${abs.toFixed(2)}`;
}

export function asSharePct(pct) {
  const n = Number(pct);
  if (!Number.isFinite(n)) return '—';
  return Math.abs(n) >= 100 ? `${n.toFixed(0)}%` : `${n.toFixed(1)}%`;
}

export function formatPercent(value) {
  const n = Number(value);
  if (!Number.isFinite(n)) return '—';
  const pct = n > 0 && n <= 1 ? n * 100 : n;
  return asSharePct(pct);
}

export function compactNumber(n) {
  const value = Number(n);
  if (!Number.isFinite(value)) return '—';
  const abs = Math.abs(value);
  const sign = value < 0 ? '-' : '';
  if (abs >= 1_000_000_000) return `${sign}${(abs / 1_000_000_000).toFixed(1)}B`;
  if (abs >= 1_000_000) return `${sign}${(abs / 1_000_000).toFixed(0)}M`;
  if (abs >= 1_000) return `${sign}${(abs / 1_000).toFixed(1)}K`;
  if (abs >= 100 || abs === Math.round(abs)) return `${sign}${abs.toFixed(0)}`;
  return `${sign}${abs.toFixed(1)}`;
}

/**
 * @param {number} value raw numeric value
 * @param {string} formatOrColumn explicit format or column key for detection
 * @param {{ tooltip?: boolean }} [options]
 */
export function formatMetricValue(value, formatOrColumn, { tooltip = false } = {}) {
  const n = Number(value);
  if (!Number.isFinite(n)) {
    return tooltip ? NOT_AVAILABLE : DASH;
  }

  const format = KNOWN_FORMATS.has(formatOrColumn)
    ? formatOrColumn
    : detectFormat(formatOrColumn);

  switch (format) {
    case 'currency':
      return asCurrency(n);
    case 'percent':
      return formatPercent(n);
    case 'number':
      return compactNumber(n);
    case 'text':
    case 'date':
    default:
      if (Number.isInteger(n) || Math.abs(n) >= 100) {
        return compactNumber(n);
      }
      return String(n);
  }
}

/** Resolve chart value format from spec metadata. */
export function resolveChartValueFormat(spec) {
  if (spec?.valueFormat && KNOWN_FORMATS.has(spec.valueFormat)) {
    return spec.valueFormat;
  }
  const columnKey = spec?.valueKey || spec?.yKey || '';
  const detected = detectFormat(columnKey);
  return detected === 'text' || detected === 'date' ? 'number' : detected;
}
